compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/NameCollision.hs

```diff
@@ -0,0 +1,165 @@
+-- Copyright (c) 2019 The DAML Authors. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+module DA.Daml.LF.TypeChecker.NameCollision
+    ( checkModule
+    ) where
+
+import DA.Daml.LF.Ast
+import DA.Daml.LF.TypeChecker.Env
+import DA.Daml.LF.TypeChecker.Error
+import Data.Maybe
+import Control.Monad.Extra
+import qualified Data.Map.Strict as M
+import qualified Data.Text as T
+import qualified Control.Monad.State.Strict as S
+
+-- | The various names we wish to track within a package.
+-- This type separates all the different kinds of names
+-- out nicely, and preserves case sensitivity, so the
+-- names are easy to display in error messages. To get
+-- the corresponding case insensitive fully resolved name,
+-- see 'FRName'.
+data Name
+    = NModule ModuleName
+    | NRecordType ModuleName TypeConName
+    | NVariantType ModuleName TypeConName
+    | NEnumType ModuleName TypeConName
+    | NVariantCon ModuleName TypeConName VariantConName
+    | NEnumCon ModuleName TypeConName VariantConName
+    | NField ModuleName TypeConName FieldName
```

Comment:
Can field and choice names actually collide with anything?

Reply:
I don't think so, at least not from DAML, but it's in the spec (unlike fully resolved values, for some reason).